ksqldb-engine/src/main/java/io/confluent/ksql/planner/JoinTree.java

```diff
@@ -262,9 +264,9 @@ public String debugString(final int indent) {
       return Streams
           .concat(
               left.viableKeyColumns().stream(),
-              Stream.of(info.getLeftJoinExpression()),
+              emptyIfNonKeyColumn(info.getLeftJoinExpression(), info.getLeftSource()),
```

Comment:
I don't think we should be updating `viableKeyColumns()` at all. I think we should change the logic in `buildJoinKey()` to have a separate pathway for foreign key joins, since the key columns for FK joins are always simply the key columns of the left source.

However, we do need to update `joinEquivalenceSet()` to reflect that the join expressions (definitely at least the left one) should not be added to the equivalence set in the case of FK joins. Otherwise the detection of PK vs FK joins downstream (for n-way joins) will be messed up, since the logic will mistakenly think that the output is partitioned on the foreign key column, which it's not.

Reply:
I looked at the changes, but I don't see the difference with ignoring FK expressions between `viableKeyColumns()` and `joinEquivalenceSet()`. The `joinEquivalenceSet()` returns columns and expressions. Then, the `viableKeyColumns()` filters the result to keep only columns. I just added the extra filter to remove FK columns. If I update the `joinEquivalenceSet()` to not add FK columns, then it won't return the expressions ,which is what `viableKeyColumns()` is filtering out.

I thought `viableKeyColumns()` was the method to do it 'cause it returns keys. Not necessary a foreign-key, but a primary key. A key that can be used to repartition only (perhaps we can call it viablePrimaryKeyColumns?). Also, it was easier to check if the column was a key in the schema. In `buildJoinKey`, I would need to look for the source that matches the alias + table name, then get the schema. Not a big change, but an extra thing to do.

Reply:
Actually, I just found out the issue with this code. The `EndToEndIntegrationTest.shouldSupportDroppingAndRecreatingJoinQuery` was failing because this query:
```
final String createStreamStatement = format(
        "create stream cart_event_product as "
        + "select pv.userid, pv.pageid, u.gender "
        + "from %s pv left join %s u on pv.userid=u.userid;",
        PAGE_VIEW_STREAM, USER_TABLE);
```

It is a stream-table join. The `pv.userid` is in the join expression and in the projection. The `u.userid` is not in the projection. The `pv.userid` is a non-key in the stream, but the plan treats the `pv.userid` as the key because it has the same type as the `u.userid` which is a key. In my code, I am removing `pv.userid` because it is not a key, and `u.userid` is also removed because it is not in the projection, so the test fails with an error that the key must be in the projection.

I fixed with your suggestion. I am not touching the `viablekeyColumns` anymore. I just added a filter in the `buildJoinKey` to ignore the fk key found in the join expression.